File: 11.cpp
//
// Created by linkinpark213 on 5/6/19.
//

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int maxArea(vector<int> &height) {
        int max_volume = 0;
        int volume;
        for (int i = 0; i < height.size(); i++) {
            if (height[i] > 0)
                for (int j = i + (max_volume / height[i]); j < height.size(); j++) {
                    volume = (j - i) * min(height[i], height[j]);
                    if (volume > max_volume) {
                        max_volume = volume;
                    }
                }
        }
        return max_volume;
    }
};

int main() {
    Solution solution;
    int nums[] = {1, 8, 6, 2, 5, 4, 8, 3, 7};
    vector<int> height(nums, nums + sizeof(nums) / sizeof(int));
    cout << solution.maxArea(height);
}

File: pysrc/11.py
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        maxA = 0
        while l < r:
            maxA = max(maxA, (r - l) * min(height[l], height[r]))
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return maxA


if __name__ == '__main__':
    solution = Solution()
    print(solution.maxArea(height=[1, 8, 6, 2, 5, 4, 8, 3, 7]))
